### PR TITLE
C++: Add PreprocBlock.qll library

### PR DESCRIPTION
**Pull request by geoffw0**
The `PreprocBlock.qll` library offers a view of `#if`, `#elif`, `#else` and similar directives (and the code they contain) as a tree.  That is, there's a parent-child relationship between blocks (identified by their opening directive), as well as the ability to tell what include statements and macro definitions are contained within each block.

---

The `preprocblock.expected` test is a good place to start if the above isn't crystal clear.

This is actually an old library, but until now it's not been available in the codeql repo.

### CODE REVIEW

File: cpp/ql/lib/change-notes/2024-01-30-preproc-block.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+---
+category: minorAnalysis
```
Comment (MathiasVP) [inline cpp/ql/lib/change-notes/2024-01-30-preproc-block.md:2]: I think this should actually be categorized as `feature`: https://github.com/github/codeql/blob/main/docs/change-notes.md#library-pack-change-categories

File: cpp/ql/lib/semmle/code/cpp/headers/PreprocBlock.qll [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+  /**
+   * Gets a textual representation of this element.
+   */
+  string toString() { result = mkElement(this).toString() }
+
+  /**
+   * Gets the file this `PreprocessorBlock` is located in.
+   */
+  cached
```
Comment (MathiasVP) [inline cpp/ql/lib/semmle/code/cpp/headers/PreprocBlock.qll:73]: I'm slightly worried about all these `cached` annotations on all predicates. Other than possibly `getPreprocLineFromIndex`, none of these predicates look very expensive

**Review by MathiasVP — COMMENTED** [review]
A few comments. I'd like to see two things fixed before we merge this:
- I think we should strip away all the `cached` and only re-introduce them if we actually end up seeing expensive re-evaluation. 
- I think we should fix all the implicit `this` warnings

But other than that this LGTM!

File: cpp/ql/lib/change-notes/2024-01-30-preproc-block.md (hunk 1 above)
Comment (geoffw0) [reply cpp/ql/lib/change-notes/2024-01-30-preproc-block.md]: Good point.  Updated.

**Comment by geoffw0** [comment]
Thanks for the comments.  I agree and I've addressed the issues with implicit this and the change note.  Still to remove the `cached` annotations, I want to do that somewhat carefully to make sure performance isn't impacted.

File: cpp/ql/lib/semmle/code/cpp/headers/PreprocBlock.qll (hunk 2 above)
Comment (geoffw0) [reply cpp/ql/lib/semmle/code/cpp/headers/PreprocBlock.qll]: Removing the `cached` annotations appears to _improve_ performance on some artificial code, and makes no consistent difference in the `RedundantInclude.ql` and `UselessInclude.ql` queries.  I'm certainly happier without annotations littered around, so I've removed them as suggested.  :+1:

**Review by MathiasVP — APPROVED** [review]
LGTM! I agree a DCA run isn't needed since nothing actually imports this library yet

**Pull request merged**